java/arcs/android/storage/service/StorageServiceNgImpl.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.android.storage.service
+
+import android.os.Parcel
+import androidx.annotation.VisibleForTesting
+import arcs.android.storage.readStoreOptions
+import arcs.core.storage.ActiveStore
+import arcs.core.storage.DriverFactory
+import arcs.core.storage.StorageKey
+import arcs.core.storage.UntypedActiveStore
+import arcs.core.storage.UntypedProxyMessage
+import arcs.core.storage.WriteBackProvider
+import java.util.concurrent.ConcurrentHashMap
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+/**
+ * Implementation of the [IStorageServiceNg] AIDL interface. Responsible for forwarding messages
+ * to [ActiveStore]s and back again.
+ */
+class StorageServiceNgImpl(
+  private val scope: CoroutineScope,
+  private val driverFactory: DriverFactory,
+  private val writeBackProvider: WriteBackProvider,
+  private val devToolsProxy: DevToolsProxyImpl?,
+  /** Callback to trigger when a proxy message has been received and sent to the store. */
+  private val onProxyMessageCallback: suspend (StorageKey, UntypedProxyMessage) -> Unit
+) : IStorageServiceNg.Stub() {
+  // TODO(b/162747024): Replace this with an LruCache so its size doesn't grow unbounded.
+  @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+  val stores = ConcurrentHashMap<StorageKey, UntypedActiveStore>()
+
+  private val stats = BindingContextStatsImpl()
+
+  override fun openStorageChannel(
+    parcelableStoreOptions: ByteArray,
+    channelCallback: IStorageChannelCallback,
+    messageCallback: IMessageCallback
+  ) {
+    val storeOptions = with(Parcel.obtain()) {
+      unmarshall(parcelableStoreOptions, 0, parcelableStoreOptions.size)
+      setDataPosition(0)
+      readStoreOptions()!!
```

Comment:
I think we always prefer `checkNotNull` + descriptive message to `!!`

Can this parcel reading code be defined has a helper/extension method somewhere?

Reply:
Done

Reply:
awesome!